### PR TITLE
Add polars compatibility

### PR DESCRIPTION
**Pull request by psmyth94**
Hey there,

I've just finished adding support to convert and format to `polars.DataFrame`. This was in response to the open issue about integrating Polars [#3334](https://github.com/huggingface/datasets/issues/3334). Datasets can be switched to Polars format via `Dataset.set_format("polars")`. I've also included `to_polars` and `from_polars`. All polars functions are checked via config.POLARS_AVAILABLE. 

A few notes:
This only supports `DataFrames` and not `LazyFrames`. This probably could be integrated fairly easily via `is_lazy` args in `set_format`, and `to_polars`.

Let me know your feedbacks.

### CODE REVIEW

**Comment by psmyth94** [comment]
Hi @lhoestq, thanks for pointing out the missing `to_polars` method.

I see your point about `table.py` so I removed them.

I also added tests in `test_arrow_dataset.py`, `test_dataset_dict.py`, and `test_formatting.py`. Let me know if I am missing any.

File: src/datasets/arrow_dataset.py [hunk 1]
```diff
@@ -3464,6 +3511,11 @@ def init_buffer_and_writer():
                                 writer.write_row(example)
                             elif isinstance(example, pd.DataFrame):
                                 writer.write_row(pa.Table.from_pandas(example))
+                            elif config.POLARS_AVAILABLE:
+                                import polars as pl
+
+                                if isinstance(example, pl.DataFrame):
+                                    writer.write_table(example.to_arrow())
```
Comment (lhoestq) [inline src/datasets/arrow_dataset.py:3518]: Here you should write_row instead of write_table

File: src/datasets/arrow_dataset.py [hunk 2]
```diff
@@ -3464,6 +3511,11 @@ def init_buffer_and_writer():
                                 writer.write_row(example)
                             elif isinstance(example, pd.DataFrame):
                                 writer.write_row(pa.Table.from_pandas(example))
+                            elif config.POLARS_AVAILABLE:
+                                import polars as pl
+
+                                if isinstance(example, pl.DataFrame):
```
Comment (lhoestq) [inline src/datasets/arrow_dataset.py:3517]: If the example is not a `pl.DataFrame` then in your current code the line `writer.write(example)` is not called. you can group `config.POLARS_AVAILABLE` and `isinstance(example, pl.DataFrame)` in the same `if` statement to fix this instead. This way the code would continue in the `else` if the example is not a polars DataFrame.

You might have to check for `"polars" in sys.modules` in the `if`, and also add this line before `with contextlib.ExitStack() as stack`:

```python
    if config.POLARS_AVAILABLE and "polars" in sys.modules:
        import polars as pl
```

File: src/datasets/arrow_dataset.py [hunk 3]
```diff
@@ -3497,6 +3549,11 @@ def init_buffer_and_writer():
                                 writer.write_table(batch)
                             elif isinstance(batch, pd.DataFrame):
                                 writer.write_table(pa.Table.from_pandas(batch))
+                            elif config.POLARS_AVAILABLE:
+                                import polars as pl
+
+                                if isinstance(batch, pl.DataFrame):
```
Comment (lhoestq) [inline src/datasets/arrow_dataset.py:3555]: same here

**Review by lhoestq — COMMENTED** [review]
Thanks ! Can you addd polars to the test dependencies in setup.py ? This way your tests will be run in the CI

I also added a few more comments:

File: src/datasets/arrow_dataset.py (hunk 1 above)
Comment (psmyth94) [reply src/datasets/arrow_dataset.py]: fixed

File: src/datasets/arrow_dataset.py [hunk 4]
```diff
@@ -3464,6 +3515,8 @@ def init_buffer_and_writer():
                                 writer.write_row(example)
                             elif isinstance(example, pd.DataFrame):
                                 writer.write_row(pa.Table.from_pandas(example))
+                            elif config.POLARS_AVAILABLE and isinstance(example, pl.DataFrame):
```
Comment (lhoestq) [inline src/datasets/arrow_dataset.py:3518]:
```suggestion
                            elif config.POLARS_AVAILABLE and "polars" in sys.modules and isinstance(example, pl.DataFrame):
```

File: src/datasets/arrow_dataset.py [hunk 5]
```diff
@@ -3497,6 +3550,8 @@ def init_buffer_and_writer():
                                 writer.write_table(batch)
                             elif isinstance(batch, pd.DataFrame):
                                 writer.write_table(pa.Table.from_pandas(batch))
+                            elif config.POLARS_AVAILABLE and isinstance(batch, pl.DataFrame):
```
Comment (lhoestq) [inline src/datasets/arrow_dataset.py:3553]:
```suggestion
                            elif config.POLARS_AVAILABLE and "polars" in sys.modules and isinstance(batch, pl.DataFrame):
```

**Review by lhoestq — COMMENTED** [review]
This should fix the CI :)

File: src/datasets/arrow_dataset.py (hunk 4 above)
Comment (psmyth94) [reply src/datasets/arrow_dataset.py]: Thanks 🤗

File: setup.py [hunk 6]
```diff
@@ -240,6 +241,7 @@
     ],
     "tensorflow_gpu": ["tensorflow-gpu>=2.2.0,!=2.6.0,!=2.6.1"],
     "torch": ["torch"],
+    "polars": ["polars"],
```
Comment (lhoestq) [inline setup.py:244]: no need to add it here
```suggestion
    "polars": ["polars"],
```

Comment (psmyth94) [reply setup.py]: Noted

File: tests/test_arrow_dataset.py [hunk 7]
```diff
@@ -2365,6 +2382,39 @@ def test_to_pandas(self, in_memory):
...
+
+                # With index mapping
+                with dset.select([1, 0, 3]) as dset:
+                    dset_to_polars = dset.to_polars()
+                    self.assertIsInstance(dset_to_polars, pl.DataFrame)
+                    self.assertEqual(len(dset_to_polars), 3)
+                    self.assertListEqual(sorted(dset_to_polars.columns), sorted(dset.column_names))
+
+                    for col_name in dset.column_names:
+                        self.assertEqual(len(dset_to_polars[col_name]), dset.num_rows)
```
Comment (lhoestq) [inline tests/test_arrow_dataset.py:2416]: Before closing the temporary directory it's important to delete all the references to opened files on Windows

Can you try adding this ? Hopefully this fixes the Windows CI
```suggestion
                        self.assertEqual(len(dset_to_polars[col_name]), dset.num_rows)
                    del dset_to_polars
```

File: tests/test_arrow_dataset.py [hunk 8]
```diff
@@ -2365,6 +2382,39 @@ def test_to_pandas(self, in_memory):
...
+            # Batched
+            with self._create_dummy_dataset(in_memory, tmp_dir, multiple_columns=True) as dset:
+                batch_size = dset.num_rows - 1
+                to_polars_generator = dset.to_polars(batched=True, batch_size=batch_size)
+
+                for batch in to_polars_generator:
+                    self.assertIsInstance(batch, pl.DataFrame)
+                    self.assertListEqual(sorted(batch.columns), sorted(dset.column_names))
+                    for col_name in dset.column_names:
+                        self.assertLessEqual(len(batch[col_name]), batch_size)
```
[the remark this answers is not in the document]
Comment (psmyth94) [reply tests/test_arrow_dataset.py:2397]: Still the same error unfortunately.
`FAILED tests/test_arrow_dataset.py::BaseDatasetTest::test_to_polars_on_disk - NotADirectoryError: [WinError 267] The directory name is invalid: 'C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\tmp4nfzb9b7\\dataset0.arrow'
`

[the remark this answers is not in the document]
Comment (psmyth94) [reply tests/test_arrow_dataset.py:2397]: The culprit is to_polars_generator

[the remark this answers is not in the document]
Comment (psmyth94) [reply tests/test_arrow_dataset.py:2397]: Should be fixed now

**Comment by psmyth94** [comment]
duckdb index files were deleted yesterday in dataset_with_script@ref/convert/parquet so I changed the hash to reflect the new SHA.

**Review by lhoestq — APPROVED** [review]
Great ! Merging now, congrats ! 🚀

**Comment by lhoestq** [comment]
I'm so excited I tweeted about it: https://x.com/qlhoest/status/1766135995513082086?s=20 I hope it's fine !

**Comment by psmyth94** [comment]
Thanks @lhoestq for the support and totally fine with the share! Happy to see people excited for this 😃

**Pull request merged**